jujugui/static/gui/src/app/store/env/controller-api.js

```diff
@@ -0,0 +1,683 @@
+        if (callback) {
+          callback({error: err});
+        }
+        return;
+      }
+      if (this.ws.readyState !== 1) {
+        console.log(
+          'Websocket is not open, dropping request. ' +
+          'readyState: ' + this.ws.readyState, op);
+        return;
+      }
+      op.version = version;
+      var tid = this._counter += 1;
+      if (callback) {
+        this._txn_callbacks[tid] = callback;
+      }
+      op['request-id'] = tid;
+      if (!op.params) {
+        op.params = {};
+      }
+      var msg = JSON.stringify(op);
+      this.ws.send(msg);
+    },
+
+    /**
+     * React to the results of sending a login message to the server.
+     *
+     * @method handleLogin
+     * @param {Object} data The response returned by the server.
+     * @return {undefined} Nothing.
+     */
+    handleLogin: function(data) {
+      this.pendingLoginResponse = false;
+      this.userIsAuthenticated = !data.error;
+      if (this.userIsAuthenticated) {
+        var response = data.response;
+        // If login succeeded store the facades and user information, and
+        // retrieve model info.
+        var facadeList = response.facades || [];
+        var facades = facadeList.reduce(function(previous, current) {
+          previous[current.name] = current.versions;
+          return previous;
+        }, {});
+        this.set('facades', facades);
+        var userInfo = response['user-info'];
+        this.set('readOnly', !!userInfo['read-only']);
+        this.set('serverTag', response['server-tag']);
+        // Start pinging the server.
+        // XXX frankban: this is only required as a temporary workaround to
+        // prevent Apache to disconnect the WebSocket in the embedded Juju.
+        if (!this._pinger) {
+          this._pinger = setInterval(
+            this.ping.bind(this), PING_INTERVAL * 1000);
+        }
+        // Clean up for log out text.
+        this.failedAuthentication = false;
+      } else {
+        // If the credentials were rejected remove them.
+        this.setCredentials(null);
+        this.failedAuthentication = true;
+      }
+      this.fire('login', {data: {
+        result: this.userIsAuthenticated,
+        error: data.error || null
+      }});
+    },
+
+    /**
+      Return a version for the given facade name which is supported by the
+      current Juju controller. If a version is provided, return the version
+      number itself if supported, or null if that specific version is not
+      served by the controller. Otherwise, if no version is specified, return
+      the most recent supported version or null if the facade is not found.
+
+      @method findFacadeVersion
+      @param {String} name The facade name (for instance "Application").
+      @param {Int} version The optional facade version (for instance 1 or 2).
+      @return {Int} The facade version or null if facade is not supported.
+    */
+    findFacadeVersion: function(name, version) {
```

Comment:
as mentioned above, there are a number of methods that need to be pulled out to be shared